README.md
# xylem

Library to interact with CarbonCulture's datastore


## Installation

Installation using pip::

    pip install https://github.com/CarbonCulture/xylem/archive/v0.4.11.zip

## Usage

You'll need an API key and an access name (like a username). Please contact
developer@carbonculture.net to request a key and access name.

### CarbonCulture's entities

CarbonCulture's datastreams are divided into entity types, which consist of
**Communities**, **Places** and **Meters**.
A **Community** is a group of people, who may or
may not exist in the same physical location, a **Place** is a physical entity
around which you could draw a line (if you had enough chalk and lots of
ladders). A **Meter** is really an energy data-source rather than necessarily
being a true physical metering point, and could be derived from one, two (or
more) input datastreams (e.g. real-time data and a backup half-hourly feed).

You are able to request a section of data for a particular entity on the
CarbonCulture platform, over a time range, at different resolutions, in
different units, and with other helpful metadata -- depending on your access
level and the availability of these data.

In order to access these data, you will need to know the unique identifier (ID)
of the entity, and this can be found in the URL of the page on
carbonculture.net that represents this entity, for example UK Parliament (the
Community) has an ID of 2:

https://platform.carbonculture.net/communities/uk-parliament/2/

Cardiff Castle's ID is 947:

https://platform.carbonculture.net/places/cardiff-castle/947/

Note that a **Place** or **Meter** may have the same ID as a **Community**,
but **Places**
and **Meters** never share IDs (this is because internally Places and Meters are
both represented as the same data structure - a physical '**Asset**').

### Where/how to find data

CarbonCulture's datastreams ('Channels') are named in a similar way to these
URLs, so for example if you want to access UK Parliament's electricity data,
you would need the Channel called `communities.2.elec`, while Cardiff Castle's
gas data will be at `places.947.gas`.

In order to discover what utilities are accessible to you for an entity, you
can use the subject module, so let's look at an example of connecting and
discovering:

```
In [1]: from xylem.connection import Connection

In [2]: xc = Connection('YOUR API USER NAME HERE', 'YOUR API KEY HERE')

In [3]: from xylem.subjects import discover_available_resources

In [4]: discover_available_resources(xc, 2, 'communities')
Out[4]:
{u'elec': {u'data_type': u'',
  u'description': u'All electricity used by the UK Parliament community',
  u'earliest_timestamp': u'2013-07-17T00:00:00',
  u'gui_max': 5500.0,
  u'latest_timestamp': u'2015-05-12T00:00:00',
  u'licence': u'None',
  u'meta': u'',
  u'permissions': [u'read_history_full'],
  u'quality_assurance': [],
  u'resource_uri': u'/api/v1/channel/communities.2.elec',
  u'slug': u'communities.2.elec',
  u'stats': None,
  u'store_history': False,
  u'unit': u'kWh',
  u'value_type': u'accum',
  u'values': []},

...

```

There's quite a lot of info in this result, but the pertinent points are the
result keys:

```
In [5]: resources = discover_available_resources(xc, 2, 'communities')

In [6]: resources.keys()
Out[6]: [u'heat', u'elec', u'gas', u'energy']

```

This tells us what utilities there are available (i.e. configured), though this
is no guarantee that there will be data available -- to check this, inspect the
`earliest_timestamp` and `latest_timestamp` within each dictionary item.

### Fetching data

Once you know what you can see about an entity, it is easy to get data for a
range of time:

```
In [7]: from datetime import datetime, timedelta

In [8]: from pytz import utc

In [9]: earliest = datetime(2015, 01, 01, 0, 0, 0, 0, utc)

In [10]: latest = earliest + timedelta(days=7)

In [11]: values = xc.read_channel_values('communities.2.elec', earliest, latest)

```

`values` now contains a `dict` keyed by timestamp, with each value being a
further `dict` of values given in `kWh` (this Channel's default units as shown
in the earlier metadata request):

```
In [14]: values[earliest]
Out[14]: {u'kWh': 0.0}

In [15]: values[latest]
Out[15]: {u'kWh': 521617.70000000007}

```

### Transforming data

In the request above, the first value (`earliest`) is 0 - this is because this
data is given in a cumulative format ('`accum`', the default on this Channel),
with the origin set to the earliest point requested. If you want to have these
data as usage, you can simply subtract one from the other, or request it as
usage:

```

In [19]: values[latest]['kWh'] - values[latest - timedelta(minutes=30)]['kWh']
Out[19]: 1331.6999999999534

In [20]: values = xc.read_channel_values('communities.2.elec', earliest, latest, value_type='usage')

In [21]: values[latest]
Out[21]: {u'kWh': 1331.6999999999534}

```

Note that the `earliest` timestamp will not be present in a usage data request
since it is still treated as the data origin, and there is nothing to subtract
it from to give a difference (usage) value.

```

In [22]: values[earliest]
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-22-90c56dbf46f1> in <module>()
----> 1 values[earliest]

KeyError: datetime.datetime(2015, 1, 1, 0, 0, tzinfo=<UTC>)

```

You can also request energy data transformed into kilograms of carbon
(equivalent) (kgCO2e), and in pence (i.e. GBP). The advantage of asking this
question of the dataserver rather than applying a factor to the resulting kWh
values is that the dataserver knows what each of the Channel's factors are for
these transformations over time, and per component input Channel. For example,
`communities.2.energy` is actually just an aggregation of `communities.2.elec`
and `communities.2.gas`, but gas and electricity have different costs and
different carbon impacts, and these change over time. The transformation to the
input energy use data over time is applied at the lowest level possible so that
it is as close to the truth as possible. This also applies to e.g. an
electricity Channel which may be an aggregate of different Meters, and so on.

Here's how that works:

```

In [27]: values = xc.read_channel_values('communities.2.energy', earliest, latest, units='kWh,pence,kgCO2e')

In [28]: values[latest]
Out[28]:
{u'kWh': 1226567.4950116458,
 u'kgCO2e': 380901.7477941546,
  u'pence': 8488058.28979018}

```

You can also change the resolution of the data, which is half-hourly by
default. In some cases you may have access to higher-resolution data but
generally you will only be able to reduce this, so for example if you want one
point per day over a month:

```

In [32]: values = xc.read_channel_values('communities.2.energy', earliest, latest, resolution=60 * 60 * 24, value_type='usage')

In [33]: values.values()
Out[33]:
[{u'kWh': 196027.87567138532},
 {u'kWh': 148819.09393853648},

 ...

 ```
 
 **NB: Some requests may take a long time to process. If you are experiencing
 multiple time outs or error responses, please let us know
 (developer@carbonculture.net).**


CHANGES.txt
Change-log for xylem.

0.4.12
------

Change expected response from PATCH to include 204

0.4.11
------

Add overwrite param to channel values write


0.4.10
------

Bugfix for read values with single unit


0.4.9
-----

Simple read values method in connection


0.4.8
-----

Function to get minimum presence of data over a date range/channel range


0.4.7
-----

Allow permission assignment (groups) for datausers.


0.4.6
-----

Fix URL encoding


0.4.5
-----

Minor bugfix

0.4.4
-----

Minor bugfix


0.4.3
-----

List datausers now re.


0.4.2
-----

Added list and create datausers.


0.4.1
-----

Added create_channels helper
Added assign_permissions_for_user_on_channel


0.4.0
-----

Introduced read_channel_latest_n_values. First read-out of values (discrete only
for now)


tests/test_api.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import json

from iso8601.iso8601 import Utc

from datetime import datetime
from unittest import TestCase

import httpretty

from xylem.connection import Connection, ROOT
from xylem.subjects import (
    discover_available_resources, minimum_data_presence_for_range
)


BASIC_RESOURCES_AVAILABLE = """
{
    "channel": {
        "list_endpoint": "/api/v1/channel/",
        "schema": "/api/v1/channel/schema/"
    }
}
"""


PLACE_BASIC_UTILS = """
{
    "meta": {
        "limit": 20,
        "next": null,
        "offset": 0,
        "previous": null,
        "total_count": 2
    },
    "objects": [
        {
            "description": "Total grid electricity consumed at SOMEPLACE",
            "earliest_timestamp": "2000-07-27T01:01:01",
            "gui_max": 500,
            "latest_timestamp": "2013-12-09T10:51:01.719301",
            "licence": "A Licence",
            "permissions": [ "read_history_hh" ],
            "resource_uri": "/api/v1/channel/places.N.elec",
            "slug": "places.N.elec",
            "stats": null,
            "store_history": false,
            "unit": "kWh",
            "value_type": "accum",
            "values": [ ]
        },
        {
            "description": "Total mains gas consumed at SOMEPLACE",
            "earliest_timestamp": "2000-07-27T01:01:01",
            "gui_max": 500,
            "latest_timestamp": "2013-12-09T10:51:01.719301",
            "licence": "A Licence",
            "permissions": [ "read_history_hh" ],
            "resource_uri": "/api/v1/channel/places.N.gas",
            "slug": "places.N.gas",
            "stats": null,
            "store_history": false,
            "unit": "m3",
            "value_type": "accum",
            "values": [ ]
        }
    ]
}
"""


VALUES_RESPONSE_MULTI_UNIT = {
    "meta": {
        "timezone": "UTC",
        "units": [
            "kWh",
            "pence",
            "kgCO2e"
        ],
        "values__earliest": "2014-12-01T00:00:00+00:00",
        "values__latest": "2014-12-02T00:00:00+00:00"
    },
    "objects": [
        {
            "resource_uri": "/api/v1/channel/a.b.c",
            "slug": "a.b.c",
            "store_history": False,
            "unit": "kWh",
            "value_type": "accum",
            "values": [
                [
                    "2014-12-01T00:00:00+00:00",
                    [
                        0,
                        0,
                        0
                    ]
                ],
                [
                    "2014-12-01T00:30:00+00:00",
                    [
                        3557.8727900740487,
                        30241.918715629377,
                        1866.531223128648
                    ]
                ],
            ]
        }
    ]
}


class XylemTestCase(TestCase):

    @httpretty.activate
    def test_resource_discovery(self):
        """Simple test to make sure dict ends up in the right format."""
        httpretty.register_uri(
            httpretty.GET, "{0}/api/v1".format(ROOT),
            body=BASIC_RESOURCES_AVAILABLE, content_type="application/json"
        )
        xc = Connection('fake', 'fake')
        httpretty.register_uri(
            httpretty.GET, xc.services['channel'],
            body=PLACE_BASIC_UTILS, content_type="application/json"
        )
        resources = discover_available_resources(xc, 'N')

        self.assertEqual(sorted(resources.keys()), sorted(['elec', 'gas']))

    @httpretty.activate
    def test_read_values_multi_units(self):
        """Should get a dict keyed by timestamp, of dicts keyed by unit."""
        httpretty.register_uri(
            httpretty.GET, "{0}/api/v1".format(ROOT),
            body=BASIC_RESOURCES_AVAILABLE, content_type="application/json"
        )
        xc = Connection('fake', 'fake')
        httpretty.register_uri(
            httpretty.GET, xc.services['channel'],
            body=json.dumps(VALUES_RESPONSE_MULTI_UNIT),
            content_type="application/json"
        )
        earliest = datetime(2014, 12, 1, 0, 0, 0, 0, Utc())
        latest = datetime(2014, 12, 1, 0, 30, 0, 0, Utc())
        result = xc.read_channel_values(
            'a.b.c', earliest, latest, units=['kWh', 'pence', 'kgCO2e']
        )
        self.assertTrue(earliest in result)
        self.assertTrue(latest in result)
        self.assertTrue('kWh' in result[earliest])
        self.assertTrue('pence' in result[earliest])
        self.assertTrue('kgCO2e' in result[earliest])


class QATests(TestCase):

    @httpretty.activate
    def test_min_presence_check(self):
        httpretty.register_uri(
            httpretty.GET, "{0}/api/v1".format(ROOT),
            body=BASIC_RESOURCES_AVAILABLE, content_type="application/json"
        )
        resp = {
            "meta": {
                "total_count": 2,
                "qa_only": True,
                "quality_assurance": [
                    "presence"
                ],
                "values__earliest": "2014-12-01T00:00:00Z",
                "values__latest": "2014-12-10T00:00:00Z"
            },
            "objects": [
                {
                    "quality_assurance": [
                        ('TS HERE', [0.3]),
                        ('TS HERE', [1]),
                    ]
                },
                {
                    "quality_assurance": [
                        ('TS HERE', [0.5]),
                        ('TS HERE', [1]),
                    ]
                }
            ]
        }
        xc = Connection('fake', 'fake')
        httpretty.register_uri(
            httpretty.GET, xc.services['channel'],
            body=json.dumps(resp), content_type="application/json"
        )

        min_presence = minimum_data_presence_for_range(
            xc, datetime.now(), datetime.now(), subject_id=1, utilities=[
                'elec', 'gas'
            ]
        )
        self.assertEqual(min_presence, 0.3)

        del resp['objects'][0]
        resp['meta']['total_count'] = 1
        httpretty.register_uri(
            httpretty.GET, xc.services['channel'],
            body=json.dumps(resp), content_type="application/json"
        )
        min_presence = minimum_data_presence_for_range(
            xc, datetime.now(), datetime.now(), slug='a.b.c'
        )
        self.assertEqual(min_presence, 0.5)


xylem/assets.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Methods for retreiving basic info about an asset's feeds."""


def discover_available_resources(conn, asset_id, asset_type='places'):
    """Return a dictionary of channels keyed by the resource slug.

    :param xylem.XylemConnector conn: The XylemConnector configured to the API
    :param int asset_id: ID of the asset to locate.
    :param str asset_type: Default: 'places'.
    :rtype dict: Resource-keyed channel info (such as {'elec': {...}})

    """
    resources = {}
    channel_root = '{0}.{1}'.format(asset_type, asset_id)
    channels = conn.list_channels(slug__startswith=channel_root)
    for slug, ch in channels.items():
        resource_slug = slug[slug.index(channel_root) + len(channel_root) + 1:]
        resources[resource_slug] = ch

    return resources




xylem/subjects.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Methods for retreiving basic info about a subject's feeds."""
from datetime import datetime
from collections import defaultdict

import pytz

from xylem.connection import HttpError


class APIError(Exception):
    pass


def discover_available_resources(conn, subject_id,
                                 subject_type_plural='places'):
    """Return a dictionary of channels keyed by the resource slug.

    :param xylem.Connection conn: The connection configured to the API
    :param int subject_id: ID of the subject to locate.
    :param str subject_type_plural: Default: 'places'.
    :rtype dict: Resource-keyed channel info (such as {'elec': {...}})

    """
    resources = {}
    channel_root = ".".join([subject_type_plural, str(subject_id)]) + '.'
    channels = conn.list_channels(slug__startswith=channel_root)
    for slug, ch in channels.items():
        resource_slug = slug[slug.index(channel_root) + len(channel_root):]
        resources[resource_slug] = ch

    return resources


def write_app_event(conn, app_slug, event_slug, event_data, subject_id,
                    subject_type_plural='communities'):
    """Write event_data to the subject's event_slug channel.

    :param xylem.Connection conn: The connection configured to the API.
    :param str app_slug: Slug of app for which this event is being written.
    :param str event_slug: Slug of event for which this data is being written
    :param dict event_data: key-value of data that describes this event.
    :param int subject_id: ID of the subject about which this event is made.
    :param str subject_type_plural: Default: 'communities'

    """
    channel_slug = ".".join([
        subject_type_plural, str(subject_id),
        'apps', app_slug, 'events', event_slug,
    ])
    now = datetime.utcnow().replace(tzinfo=pytz.utc)
    code, message = conn.write_channel_values(
        channel_slug, [(now.isoformat(), event_data)])
    if code not in [202, 204]:
        raise APIError("API Error (Code: {0}): {1}".format(code, message))


def discover_installed_apps(conn, subject_id,
                            subject_type_plural='communities'):
    """Return a list of app slugs installed, and visible, on this subject.

    :param xylem.Connection conn: The connection configured to the API
    :param int subject_id: ID of the subject to locate.
    :param str subject_type_plural: Default: 'communities'.
    :rtype dict: keys are app slugs, values are channels dicts keyed by slug
        available in that app

    Note that there may be apps installed which do not make their availability
    visible to other parties.

    """
    channel_root = ".".join([
        subject_type_plural, str(subject_id), 'apps'
    ]) + '.'
    try:
        channels = conn.list_channels(slug__startswith=channel_root)
        apps = defaultdict(dict)
        for slug, ch in channels.items():
            app_part = slug[slug.index(channel_root) + len(channel_root):]
            app_slug = app_part.split('.')[0]
            apps[app_slug][slug] = ch

        return apps
    except HttpError as e:
        raise APIError("API Error: {0}".format(e))


def minimum_data_presence_for_range(conn, earliest, latest, slug=None,
                                    subject_id=None, subject_type_plural=None,
                                    utilities=None):

    """Return minimum data presence for given subject: 0 to 1 (100%).

    Note that for subjects with real-time data, this is currently unreliable.
    Note that this represents data presence for known & configured sources.

    :param xylem.Connection conn: The connection configured to the API
    :param datetime earliest: from when to get presence value
    :param datetime latest: up to (inclusive) when to get presence value
    :param str slug: slug to get presence for, or None to use subject type/id
    :param int subject_id: ID of the subject to locate, or None if using slug
    :param str subject_type_plural: Default: 'places'
    :param list utilities: list of utilities (elec, gas, etc.) or None
    :rtype: float
    :return: a value between 0 (no data) and 1 (100% expected data present)
    :raises: APIError in the case that either the request fails or isn't 200 OK

    """
    subject_type_plural = subject_type_plural or 'places'
    slugs = []

    if slug is None:
        assert subject_id, "You must provide a subject_id if not a slug"
        subject_id = str(subject_id)
        if utilities:
            for util in utilities:
                slug = ".".join([subject_type_plural, subject_id, util])
                slugs.append(slug)
        else:
            slugs = [".".join([subject_type_plural, subject_id])]
    else:
        slugs = [slug]

    params = {
        'qa_only': True,
        'quality_assurance': 'presence',
        'values__earliest': earliest.isoformat(),
        'values__latest': latest.isoformat(),
        'slug__in': ",".join(slugs)
    }
    try:
        resp = conn.get(
            endpoint=conn.services['channel'], params=params
        )
    except HttpError as e:
        raise APIError("API Error: {}".format(e))

    if resp.status_code != 200:
        raise APIError(
            "API Error: ({}) {}".format(resp.status_code, resp.content))

    _json = resp.json()
    if _json['meta']['total_count'] != len(slugs):
        present = [x['slug'] for x in _json['objects']]
        missing = set(slugs) - set(present)
        raise APIError(
            "API Error: (403) You don't have access to channels: {}".format(
                ", ".join(missing)
            )
        )

    min_presence = 1
    for channel in _json['objects']:
        qa_presence = channel['quality_assurance']
        presence_vals = [vals[0] for ts, vals in qa_presence]
        min_presence = min(min_presence, min(presence_vals))
    return min_presence


xylem/connector.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import logging

import requests

from xylem import __version__

ROOT = 'http://nectarine.infra.carbonculture.net'

log = logging.getLogger(__name__)


class HttpError(Exception):
    pass


class XylemConnection(object):
    """Basic class configured to make requests to CarbonCulture's Data API."""

    def __init__(self, access_name, api_key, root=None, format=None):
        self.access_name = access_name
        self.api_key = api_key
        self.root = root or ROOT
        self.endpoint = '{0}/api/v1/'.format(self.root)
        self.format = format or 'application/json'

        self.headers = {
            'Authorization': 'ApiKey {0}:{1}'.format(
                self.access_name, self.api_key),
            'User-Agent': 'XylemConnection Version {0}'.format(__version__),
            'Accept': self.format,
        }
        self.services = {}
        self._discover(self._test_connection())

    def get(self, endpoint=None, params=None):
        """Make a get."""
        log.debug('Get: {0}'.format(endpoint or self.endpoint), extra=params)
        r = requests.get(
            endpoint or self.endpoint,
            params=params,
            headers=self.headers,
        )
        return r

    def _test_connection(self):
        """Ping the endpoint and check we get a 200"""
        r = self.get()
        if r.status_code != 200:
            raise HttpError(
                "Got response code {0} from {1}".format(
                    r.status_code, self.endpoint))
        return r

    def _discover(self, response=None):
        """Get a list of accessible services.

        :param request.Response response: A previously fetched response

        """
        response = response or self.get()
        available = response.json()
        for key, meta in available.items():
            self.services[key] = self.root + meta['list_endpoint']

    def list_channels(self, **kwargs):
        """Get a list of channels, maybe filtered with kwargs"""
        r = self.get(
            self.services['channel'],
            params=kwargs
        )
        content = r.json()

        channels = dict([(ch['slug'], ch) for ch in content['objects']])
        while content['meta']['next'] is not None:
            r = self.get(self.root + content['meta']['next'])
            content = r.json()
            channels.update(dict(
                [(ch['slug'], ch) for ch in content['objects']]))
        return channels


xylem/connection.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import logging
import json
import urlparse
import urllib

import iso8601
import requests

from xylem import __version__

ROOT = 'https://rhizome.carbonculture.net'
API_PREFIX = 'api/v1'

log = logging.getLogger(__name__)


class HttpError(Exception):
    pass


class APIError(Exception):
    pass


class Connection(object):
    """Basic class configured to make requests to CarbonCulture's Data API."""

    def __init__(self, access_name, api_key, root=None, format=None):
        self.access_name = access_name
        self.api_key = api_key
        self.root = root or ROOT
        self.endpoint = '/'.join([self.root, API_PREFIX])
        self.format = format or 'application/json'

        self.headers = {
            'Authorization': 'ApiKey {0}:{1}'.format(
                self.access_name, self.api_key),
            'User-Agent': 'Xylem Version {0}'.format(__version__),
            'Accept': self.format,
        }
        self.services = {}
        self._discover(self._test_connection())

    def _request(self, endpoint=None, method=None, params=None, data=None,
                 extra_headers=None):
        """Generic request, default to GET."""
        method = method or 'get'
        headers = extra_headers or {}
        headers.update(self.headers)
        log.debug(
            '{0}: {1}'.format(method, endpoint or self.endpoint),
            extra={
                'params': params,
                'data': data,
                'headers': headers,
            }
        )
        fn = getattr(requests, method)
        r = fn(
            endpoint or self.endpoint,
            params=params,
            data=data,
            headers=headers,
        )
        return r

    def get(self, endpoint=None, params=None):
        """Make a get."""
        return self._request(endpoint, params=params)

    def patch(self, endpoint=None, params=None, data=None):
        """Partial update to resource (e.g. put history or change meta)."""
        if isinstance(data, dict):
            data = json.dumps(data)
        return self._request(
            endpoint, params=params, data=data, method='patch',
            extra_headers={
                'Content-Type': 'application/json',
            }
        )

    def post(self, endpoint=None, params=None, data=None):
        """Create resource."""
        if isinstance(data, dict):
            data = json.dumps(data)
        return self._request(
            endpoint, params=params, data=data, method='post',
            extra_headers={
                'Content-Type': 'application/json',
            }
        )

    def _test_connection(self):
        """Ping the endpoint and check we get a 200"""
        r = self.get()
        if r.status_code != 200:
            raise HttpError(
                "Got response code {0} from {1}".format(
                    r.status_code, self.endpoint))
        return r

    def _discover(self, response=None):
        """Get a list of accessible services.

        :param request.Response response: A previously fetched response

        """
        response = response or self.get()
        available = response.json()
        for key, meta in available.items():
            self.services[key] = self.root + meta['list_endpoint']

    def list_channels(self, **kwargs):
        """Get a list of channels, maybe filtered with kwargs"""
        r = self.get(
            self.services['channel'],
            params=kwargs
        )
        if r.status_code == 200:
            content = r.json()

            channels = dict([(ch['slug'], ch) for ch in content['objects']])
            while content['meta']['next'] is not None:
                r = self.get(self.root + content['meta']['next'])
                content = r.json()
                channels.update(dict(
                    [(ch['slug'], ch) for ch in content['objects']]))
            return channels
        else:
            raise HttpError(
                "Got response code {0} from {1}".format(
                    r.status_code, self.endpoint))

    def write_channel_values(self, channel_slug, values, overwrite=False):
        """Write the given values to the channel identified by channel_slug.

        :param str channel_slug: Slug of channel to which data will be written
        :param list values: (timestamp, value) list to write to channel
        :param bool overwrite: Default False, set True to blat old values.
        :rtype (int, str): (status code (one of: 202, 204, 401), message)
        """
        _r = self.patch(
            self.services['channel'] + channel_slug,
            data={
                'values': values,
                'overwrite': overwrite,
            },
        )

        return (_r.status_code, _r.content)

    def read_channel_values(self, channel_slug, earliest, latest, **kwargs):
        """Read values from a given channel, between earliest and latest.

        :param str channel_slug: Slug of channel to read
        :param datetime earliest: from when to get readings (note, you may not
            get a point returned at this time if you are reading e.g. usage
            data since usage is given at the end of a period, so will be given
            at earliest + resolution)
        :param datetime latest: up to when to get readings (you should get a
            point at this time)
        :param kwargs: extra kwargs to add to the params dict
        :rtype: tuple list
        :return: [(timestamp, [values]), ...]
        :raises: APIError in the case that something is wrong with the request

        """
        # TODO: this method should support retrieving stats.
        params = {
            'slug': channel_slug,
            'values__earliest': earliest.isoformat(),
            'values__latest': latest.isoformat(),
        }
        params.update(kwargs)  # e.g. resolution, units...

        _r = self.get(
            self.services['channel'],
            params=params,
        )
        if _r.status_code == 200:
            _json = _r.json()
            ch = _json['objects'][0]
            units = _json['meta'].get('units', [ch['unit']])
            results = {
                iso8601.parse_date(x[0]): {
                    unit: x[1][ix] if len(units) > 1 else x[1] for ix, unit in enumerate(units)
                }
                for x in ch['values']
            }
            return results
        raise APIError(
            "API Error: ({}) {}".format(_r.status_code, _r.content))

    def create_channel(self, channel_data):
        """Posts to the API to make a new channel. Doesn't do existence check.

        :param dict channel_data: keys and values to make this channel.
        :rtype (int, str): (status code (one of: 201, 401), message)

        """
        _r = self.post(
            self.services['channel'],
            data=channel_data,
        )

        return (_r.status_code, _r.content)

    def create_channels(self, channel_data_list):
        """Convenience method to create several channels on one channels

        :param list channel_data_list: list of dicts
        with keys and values for the channels
        :rtype list: status codes for each of the channel creation calls.

        """
        responses = []
        for ch in channel_data_list:
            responses.append(self.create_channel(ch))
        return responses

    def read_channel_latest_n_values(self, channel_slug, n=1):
        """Retrieve the latest n points from a channel.

        :param str channel_slug: Slug of channel for which to get data.
        :param int n: defaults to 1, number of points to get.
        :rtype list: tuple list of values
        """
        _r = self.get(
            self.services['channel'],
            params={
                'slug': channel_slug,
                'values__latest_n': n,
            }
        )
        response = _r.json()
        ch = response['objects'][0]
        values = ch['values']
        if isinstance(values, dict) and 'error' in values:
            raise ValueError(values['error'])
        return [
            (iso8601.parse_date(t), v)
            for t, v in values
        ]

    def assign_permissions_for_user(self, user, permissions):
        """Assign the set of regular permissions for a user (groups).

        :param str user: User access_name.
        :param list permissions: List of codenames of permissions.
        :rtype (int, str): (status code, message)
        """
        _r = self.patch(
            urlparse.urljoin(self.services['datauser'], urllib.quote(user)),
            data={
                'permissions': permissions
            },
        )

        return (_r.status_code, _r.content)

    def assign_permissions_for_user_on_channel(
            self, user, channel_slug, permissions):
        """Assign the set of permissions for the user
        and the channel.

        :param str user: User access_name.
        :param str channel_slug: Slug of the channel.
        :param list permissions: List of codenames of permissions.
        :rtype (int, str): (status code, message)
        """
        _r = self.patch(
            self.services['channel'] + channel_slug,
            data={
                'user': user,
                'permissions': permissions
            },
        )

        return (_r.status_code, _r.content)

    def list_datausers(self):
        """Retrieves a list of the datausers.

        :rtype (int, str):
        :return: (status code, message)
        """
        _r = self.get(
            self.services['datauser']
        )
        return (_r.status_code, _r.content)

    def get_datauser(self, access_name):
        """Retrieve a datauser based on an access name.

        :rtype dict:
        :return: Dict with the information for that datauser.
        """
        _r = self.get(
            urlparse.urljoin(
                self.services['datauser'], urllib.quote(access_name))
        )
        return (_r.status_code, _r.content)

    def create_datauser(self, access_name):
        """Create datauser in Rhizome. Requires staff or admin user.

        :param str access_name: DataUser access name.
        :rtype (int, str): (status code, message).
        """
        _r = self.post(
            self.services['datauser'],
            data={"access_name": access_name},
        )

        return (_r.status_code, _r.content)

    def create_datausers(self, access_names):
        """Convenience method to create several datausers.

        :param list access_names: list of access names of users.
        with keys and values for the datausers
        :rtype list: status codes for each of the datauser creation calls.
        """
        responses = []
        for name in access_names:
            r = self.create_datauser(name)
            responses.append(r)
        return responses
